docs/tutorials/embedding/MASE.ipynb

```diff
@@ -0,0 +1,374 @@
+   "outputs": [],
+   "source": [
+    "fig, ax = plt.subplots(figsize=(10,10))\n",
+    "ax.scatter(V_hat[:25, 0], V_hat[:25, 1], c=\"blue\", label=\"Block 1\")\n",
+    "ax.scatter(V_hat[25:, 0], V_hat[25:, 1], c=\"red\", label=\"Block 2\")\n",
+    "ax.legend()\n",
+    "ax.set_title(\"Common Latent Subspace from MASE Embedding\", fontsize=20)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Recover Probability Matrices"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Since ``MultipleASE()`` stores estimates of the individual score matrices $\\hat{R}^{(i)}$ as well as the common invariant subspace $\\hat{V}$, we can recover an estimate of the probability matrix for each graph \n",
+    "$$\\hat{P}^{(i)} = V R^{(i)} V^T$$"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "R = embedder.scores_\n",
+    "R_1 = R[0]\n",
+    "R_2 = R[1]\n",
+    "\n",
+    "V_hat_T = np.transpose(V_hat)\n",
+    "P_1_hat = np.matmul(np.matmul(V_hat, R_1), V_hat_T)\n",
+    "P_2_hat = np.matmul(np.matmul(V_hat, R_2), V_hat_T)\n",
+    "\n",
+    "heatmap(P_1_hat, figsize=(7,7), title=\"Visualization of Probability Matrix 1\")\n",
+    "heatmap(P_2_hat, figsize=(7,7), title=\"Visualization of Probability Matrix 2\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Recover Original Block Probabilities"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now, we can see the clear block structure once again.  In order to obtain an estmate of the original block probabilities, we can average across the blocks."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "P1_hat_block = np.array([[np.mean(P_1_hat[:25, :25]), np.mean(P_1_hat[:25, 25:])],\n",
+    "               [np.mean(P_1_hat[25:, :25]), np.mean(P_1_hat[25:, 25:])]])\n",
+    "P2_hat_block = np.array([[np.mean(P_2_hat[:25, :25]), np.mean(P_2_hat[:25, 25:])],\n",
+    "               [np.mean(P_2_hat[25:, :25]), np.mean(P_2_hat[25:, 25:])]])\n",
+    "print(\"P1_hat_block =\\n\", P1_hat_block)\n",
+    "print(\"P2_hat_block =\\n\", P2_hat_block)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    " Although we don't get exactly the same block probabilities as we input into the SBM, it's fairly close with only 1 graph for each!"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Sampling More Graphs"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can recover an even more accurate representation of the two SBMs with more sample graphs.  Let's try with 100 graphs sampled from each.  Interestingly, this time ``MultipleASE()`` chooses to embed into 4 dimensions."
```

Comment:
serious question, note testing you: do we know why it overshoots dimensions?

Reply:
*not

Reply:
That's a good question; I'm not sure.  With the default parameters, ``MultipleASE()`` uses 2 elbows in the eigenvalues to determine the embedding dimension d.

Reply:
It could be that the two-block sbm only has 1 real elbow (between eigenvalues 2 and 3), so the second elbow has to come after that?  But I'm not sure that would explain why it doesn't overshoot with only one graph...

Reply:
So, the number of embedding dimensions that ``MultipleASE()`` chooses actually depends on the particular sampled sbms.  Out of 100 tries, with ``n_elbows=2``, I got ``d=3`` about 40 times, ``d=4`` about 45 times, and ``d=5`` about 15 times.

Reply:
Honestly I'm not sure what's going on, @bdpedigo, thoughts?

Reply:
Actually, I thought about it, and I think I have an explanation.  A k-block sbm can be represented as k clusters, where the variance of each cluster gives a measure for the within-block probability, and the distance between clusters gives a measure of the between-block probabilities.  But, you can represent k equidistant clusters (or points) in k-1 dimensions (eg. triangle in two dimensions).  Since we have a 2-block sbm, we have 2 clusters, which can be faithfully represented in 1 dimension.  Thus, ``MultipleASE()`` is identifying the one true elbow between singular values 1 and 2, and all the subsequent elbows are just noise.   This would explain why we get a different number of dimensions depending on the particular sbm.  In fact, if you look at the pairplot, in dimension 1 alone, you can resolve the two different clusters, with the height of each peak showing the within-block probability, and the distance between showing the between-block probability.

Reply:
Indeed, in the case with the single sample from each sbm, after 1000 trials, we get ``d = 3`` about 20% of the time.

Reply:
> Actually, I thought about it, and I think I have an explanation. A k-block sbm can be represented as k clusters, where the variance of each cluster gives a measure for the within-block probability, and the distance between clusters gives a measure of the between-block probabilities. 

i'm not following.

> But, you can represent k equidistant clusters (or points) in k-1 dimensions (eg. triangle in two dimensions). Since we have a 2-block sbm, we have 2 clusters, which can be faithfully represented in 1 dimension.

Well... yes, but no. There is 1-dimensional linear manifold, but we are constrained to subspace, which include origin. So unless your latent positiosn are on the ray coming from the origin, you have to use a 2-dimensional subspace. By induction holds for more dimensions.

> Thus, `MultipleASE()` is identifying the one true elbow between singular values 1 and 2,

First elbow "should" be between 2 and 3 because this where the noise ends?...

> and all the subsequent elbows are just noise. 

see above.

> This would explain why we get a different number of dimensions depending on the particular sbm. In fact, if you look at the pairplot, in dimension 1 alone, you can resolve the two different clusters, with the height of each peak showing the within-block probability, and the distance between showing the between-block probability.

you can't encode two blocks in one dimension. think of this: you need to encode 3 numbers (intrablock 1, intrablock 2, interblock). you can't do that using two points (latent position 1, latent position 2) in one dimension, uneless intterblock probability is dependent on intrablock probabilities, in which case you have a ray on 1d.

Reply:
@bdpedigo, in general, ZG2 is conservative (as in "pick too many") in number of dimensions, isn't it? this is probably the answer.

Reply:
but i guess my question deals more with "why does MASE ZG2 overshoot much more often than one graph ASE ZG2"?

Reply:
i see, it's just a max over individual ZG2's (which are already conservative), so i guess this just how the distribution of the maxes is in this problem.

Reply:
So, I was curious, and tried to vary the number of blocks in the sbm and see what happens.  When I sampled two sbms from the same block probabilities, ``MultipleASE`` only ever embeds them in 2 or 3 dimensions, even as the number of blocks increases to 100.